File: src/main/java/com/wikia/tibia/jackson/Parser.java
package com.wikia.tibia.jackson;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JavaType;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.type.CollectionType;
import com.wikia.tibia.objects.WrappedWikiObject;
import lombok.extern.slf4j.Slf4j;
import org.json.JSONObject;

import java.io.IOException;
import java.util.Collections;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

@Slf4j
public class Parser {

    private static ObjectMapper defaultObjectMapper;

    private Parser() {
        // don't instantiate this class, it has only static members
    }

    public static <T> T parse(Class<T> type, String json) {
        return parse(type, getDefaultObjectMapper(), json);
    }

    public static <T> T parse(Class<T> type, ObjectMapper mapper, String json) {
        if (json == null || json.equals("")) {
            return null;
        }

        JavaType javaType = mapper.getTypeFactory().constructType(type);
        try {
            return mapper.readValue(json, javaType);
        } catch (IOException e) {
            throw new JacksonParsingException(e);
        }
    }

    public static Object parse(String json) {
        return parse(json, getDefaultObjectMapper());
    }

    public static Object parse(String json, ObjectMapper mapper) {
        return parse(Object.class, mapper, json);
    }

    public static <T> List<T> list(Class<T> type, String json) {
        return list(type, getDefaultObjectMapper(), json);
    }

    public static <T> List<T> list(Class<T> type, ObjectMapper mapper, String json) {

        if (json == null || "".equals(json)) {
            return Collections.emptyList();
        }

        List<T> infos;
        CollectionType collectionType = mapper.getTypeFactory().constructCollectionType(List.class, type);
        try {
            infos = mapper.readValue(json, collectionType);
        } catch (IOException e) {
            throw new JacksonParsingException(e);
        }

        return infos != null
                ? infos
                : Collections.emptyList();
    }

    /**
     * Alternative implementation of {@link #list(Class, String)} where the json list is read one-by-one, which makes
     * it more robust against jsonparsingexceptions
     */
    public static <T> List<T> listOneByOne(Class<T> type, String json, long limit) {
        return listOneByOne(type, getDefaultObjectMapper(), json, limit);
    }

    private static <T> List<T> listOneByOne(Class<T> type, ObjectMapper mapper, String json, long limit) {
        if (json == null || "".equals(json)) {
            return Collections.emptyList();
        }

        limit = limit != -1
                ? limit
                : Long.MAX_VALUE;

        try {
            JsonNode jsonAsNode = mapper.readTree(json);

            if (jsonAsNode.isArray()) {
                return IntStream.range(0, jsonAsNode.size())
                        .mapToObj(jsonAsNode::get)
                        .map(jn -> {
                            try {
                                T value = mapper.treeToValue(jn, type);
                                if (value instanceof WrappedWikiObject) {
                                    ((WrappedWikiObject) value).setOriginalJson(new JSONObject(value));
                                }
                                return value;
                            } catch (JsonProcessingException e) {
                                log.error("Unable to construct object of type '{}' from the following json: {} because of error: {}", type, jn, e);
                                return null;
                            }
                        })
                        .filter(Objects::nonNull)
                        .limit(limit)
                        .collect(Collectors.toList());
            } else {
                log.error("Retrieved json is not a json array, cannot parse to list.");
                return Collections.emptyList();
            }
        } catch (IOException e) {
            throw new JacksonParsingException(e);
        }
    }

    public static String json(Object object) {
        return json(object, getDefaultObjectMapper());
    }

    public static String json(Object object, ObjectMapper mapper) {
        try {
            return mapper.writeValueAsString(object);
        } catch (JsonProcessingException e) {
            log.error("Unable to serialise object to json", e);
            return "";
        }
    }

    private static ObjectMapper getDefaultObjectMapper() {
        if (defaultObjectMapper == null) {
            defaultObjectMapper = new ObjectMapper();
            defaultObjectMapper.setSerializationInclusion(JsonInclude.Include.NON_NULL); // Value that indicates that only properties with non-null values are to be included.
        }
        return defaultObjectMapper;
    }

    private static class JacksonParsingException extends RuntimeException {
        public JacksonParsingException(Exception e) {
            super(e);
        }
    }
}

File: src/main/kotlin/com/wikia/tibia/jackson/Parser.kt
package com.wikia.tibia.jackson

import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.wikia.tibia.exceptions.JacksonParsingException
import org.slf4j.LoggerFactory
import java.io.IOException

object Parser {

    private val logger = LoggerFactory.getLogger(Parser::class.java)
    private val defaultObjectMapper = jacksonObjectMapper()
            .setSerializationInclusion(JsonInclude.Include.NON_NULL)

    fun <T> parse(type: Class<T>, json: String?): T? {
        return parse(type, defaultObjectMapper, json)
    }

    fun <T> parse(type: Class<T>, mapper: ObjectMapper, json: String?): T? {
        if (json == null || json == "") {
            return null
        }
        val javaType = mapper.typeFactory.constructType(type)
        return try {
            mapper.readValue(json, javaType)
        } catch (e: IOException) {
            throw JacksonParsingException(e)
        }
    }

    fun <T> list(type: Class<T>, json: String?): List<T> {
        return list(type, defaultObjectMapper, json)
    }

    fun <T> list(type: Class<T>, mapper: ObjectMapper, json: String?): List<T> {
        if (json == null || "" == json) {
            return emptyList()
        }
        val collectionType = mapper.typeFactory.constructCollectionType(MutableList::class.java, type)
        return try {
            mapper.readValue(json, collectionType)
        } catch (e: IOException) {
            throw JacksonParsingException(e)
        }
    }

    /**
     * Alternative implementation of list() where the json list is read one-by-one, which makes
     * it more robust against jsonparsingexceptions
     */
    fun <T : Any> listOneByOne(type: Class<T>, mapper: ObjectMapper = defaultObjectMapper, json: String?, limit: Int = Int.MAX_VALUE): List<T> {
        if (json == null || "" == json) {
            return emptyList()
        }
        return try {
            val jsonAsNode: JsonNode = mapper.readTree(json)
            if (jsonAsNode.isArray) {
                (0 until jsonAsNode.size()).asSequence()
                        .map { jsonAsNode.get(it) }
                        .mapNotNull { parseJsonToWikiObject(mapper, it, type) }
                        .take(limit)
                        .toList()
            } else {
                logger.error("Retrieved json is not a json array, cannot parse to list.")
                emptyList()
            }
        } catch (e: IOException) {
            throw JacksonParsingException(e)
        }
    }

    private fun <T : Any> parseJsonToWikiObject(mapper: ObjectMapper, it: JsonNode?, type: Class<T>): T? {
        return try {
            mapper.treeToValue(it, type)
        } catch (e: JsonProcessingException) {
            logger.error("Unable to construct object of type '$type' from the following json: '$it' because of error: $e")
            null
        }
    }

    fun <T> json(someObject: T, mapper: ObjectMapper = defaultObjectMapper): String {
        return try {
            mapper.writeValueAsString(someObject)
        } catch (e: JsonProcessingException) {
            logger.error("Unable to serialise object to json", e)
            ""
        }
    }
}